pdl-derive/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use codespan_reporting::term;
+use proc_macro2::TokenStream;
+use quote::quote;
+use std::env;
+use std::path::Path;
+use syn::parse_macro_input;
+
+fn pdl_proc_macro(path: syn::LitStr, input: syn::ItemMod) -> TokenStream {
+    // Locate the source grammar file.
+    let root = env::var("CARGO_MANIFEST_DIR").unwrap_or_else(|_| ".".into());
+    let Some(relative_path) =
+        [Path::new(&root).join(path.value()), Path::new(&root).join("src").join(path.value())]
+            .into_iter()
+            .find(|path| path.exists())
+    else {
+        return syn::Error::new(path.span(), "error: unable to find file").to_compile_error();
+    };
+
+    // Load and parse the grammar.
+    let mut sources = pdl_compiler::ast::SourceDatabase::new();
+    let relative_path = relative_path.into_os_string().into_string().unwrap();
+    let file = match pdl_compiler::parser::parse_file(&mut sources, &relative_path) {
+        Ok(file) => file,
+        Err(err) => {
+            let mut buffer = termcolor::Buffer::no_color();
+            term::emit(&mut buffer, &term::Config::default(), &sources, &err)
+                .expect("could not emit parser diagnostics");
+            return syn::Error::new(path.span(), String::from_utf8(buffer.into_inner()).unwrap())
+                .to_compile_error();
+        }
+    };
+
+    // Run the analyzer.
+    let analyzed_file = match pdl_compiler::analyzer::analyze(&file) {
+        Ok(file) => file,
+        Err(diagnostics) => {
+            let mut buffer = termcolor::Buffer::no_color();
+            diagnostics.emit(&sources, &mut buffer).expect("could not emit analyzer diagnostics");
+            return syn::Error::new(path.span(), String::from_utf8(buffer.into_inner()).unwrap())
+                .to_compile_error();
+        }
+    };
+
+    // Generate the pdl backend implementation.
+    let parser = pdl_compiler::backends::rust::generate_tokens(&sources, &analyzed_file);
+    let mod_ident = input.ident;
+    let mod_attrs = input.attrs;
+    let mod_vis = input.vis;
+    let mod_items = input.content.map(|(_, items)| items).unwrap_or(vec![]);
+
+    quote! {
+        #(#mod_attrs)*
+        #mod_vis mod #mod_ident {
+            // Generate an include_str! statement to force a dependency
+            // on the source pdl file.
+            const _SOURCE: &'static str = include_str!(#relative_path);
```

Comment:
Is this trying to emulate the [`rerun-if-changed`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-changed) option from `build.rs` scripts?

It seems the problem is [described here](https://internals.rust-lang.org/t/pre-rfc-add-a-builtin-macro-to-indicate-build-dependency-to-file/9242) — without a clear solution, apart from writing a `build.rs` file in addition to using the macro.

Reply:
I am copying the workaround used by pest:
> Currently there is no direct way for those proc macro to inform compiler about the dependency. One workaround is to use include_str! or include_bytes! macro and assign its result to a const. This isn’t ideal as the generated program doesn’t necessarily need such content, while including all the data into generated source file may slow down build time and bloat memory usage, especially when the file is large.